virtual/activations.go

```diff
@@ -382,6 +383,46 @@ func (a *activations) getServerState() (
 	return a.serverState.serverID, a.serverState.serverVersion
 }
 
+func (a *activations) close(ctx context.Context, numWorkers int) error {
+	a.Lock()
+	defer a.Unlock()
+
+	// create a channel to receive keys
+	keys := make(chan types.NamespacedActorID, len(a._actors))
```

Comment:
But why aren't you a fan of it? As you said, if we have thousands of actors, it's more performant not to spawn goroutines every time. Could you tell me what the benefit of not doing that is?

Reply:
Because it’s more error prone. You’re forced to allocate a large slice of keys upfront, if you make a small mistake you could end up missing some keys or processing some keys twice. The semantics of channels in general are very error prone and edge Casey (send after close, waiting on I closed channel forever etc)

They have their use cases, but imo in this case a semaphore is simpler and I’m happy to pay to launch a goroutine per actor since the total number of concurrent goroutines is bounded

Reply:
And this is not critical path code

Reply:
> You’re forced to allocate a large slice of keys upfront

This could be fixed by using pointers. Semaphore + goroutine spawning is more complex and less performant, so I wanted to use something other than them. But If you think using the channel is riskier, I will change it.